queasars/circuit_evaluation/mutex_primitives.py

```diff
@@ -0,0 +1,303 @@
+
+        # Check that all arguments are of the same length
+        arg_length = len(args[0])
+        for arg in args:
+            if len(arg) != arg_length:
+                raise ValueError("All arguments must be sequences of the same length!")
+
+        # This region ensures that no thread may enter if a batched execution of f is currently running.
+        # If f is not currently running it ensures that entering threads append their arguments to the batch
+        # in an orderly and threadsafe manner.
+        # To successfully enter this region, entry must be currently allowed (self._entry_lock must be available)
+        # and this thread must be given write access to the shared memory (self._variable_lock must be acquired).
+        acquired_both_locks: bool = False
+        while not acquired_both_locks:
+
+            # Try to acquire both needed locks.
+            if self._entry_lock.acquire(blocking=True):
+                if self._variable_lock.acquire(blocking=False):
+                    # Both needed locks are acquired, the thread may continue.
+                    acquired_both_locks = True
+
+                    # Remember the index at which the thread's arguments are appended to the batch.
+                    batch_index = self._batch_length
+                    # Append the thread's arguments to the batch.
+                    if self._batch_length == 0:
+                        self._batched_args = tuple((*arg,) for arg in args)
+                        self._batch_length = arg_length
+                    else:
+                        self._batched_args = tuple((*previous, *new) for previous, new in zip(self._batched_args, args))
+                        self._batch_length = self._batch_length + arg_length
+
+                    # Increase the counter to keep note of how many threads have entered so far.
+                    self._thread_counter = self._thread_counter + 1
+
+                    # The thread is now done entering. Release the locks and notify the other waiting threads that
+                    # entry may be available again.
+                    self._entry_lock.release()
+                    self._variable_lock.release()
+                    with self._external_wait_condition:
+                        self._external_wait_condition.notify_all()
+                else:
+                    # Release the acquired lock, if only one could be acquired!
+                    self._entry_lock.release()
+            # If acquiring both locks was unsuccessful, wait to be notified of entry availability.
+            if not acquired_both_locks:
+                with self._external_wait_condition:
+                    self._external_wait_condition.wait()
+
+        # Wait for the given duration so that other threads may have time to enter.
+        if self.batch_waiting_duration is not None:
+            sleep(self.batch_waiting_duration)
```

Comment:
It does block the thread. Therefore, if the threadpool contains only one thread, the batch sizes will be one, but no deadlocks should occur, as once the thread has finished waiting it will just continue to execute the batch even if no other threads have joined until then. As threadpools cannot interleave the executing thread, this behaviour would be unavoidable even without the sleep as far as i can tell, without using async or an actor approach.

Reply:
Understood, that's a bit of a pity, but so be it. Thank you!